keystorage/yubihsm2/src/test/java/tech/pegasys/signers/yubihsm2/YubiHsm2Test.java

```diff
@@ -0,0 +1,178 @@
+
+public class YubiHsm2Test {
+  @TempDir static Path hsmDataDir;
+  private static final short DEFAULT_AUTH_KEY = (short) 1;
+  private static final String DEFAULT_PASSWORD = "password";
+  private static final short SECONDARY_AUTH_KEY = (short) 2;
+  private static final String SECONDARY_PASSWORD = "password2";
+  private static List<String> expectedKeys = new ArrayList<>(10);
+
+  @BeforeAll
+  static void initYubiHsmSimulator() throws IOException, TimeoutException, InterruptedException {
+    addSecondaryAuthKey();
+    addOpaqueData();
+  }
+
+  @Test
+  public void validKeysAreFetchedSuccessfully() {
+    final YubiHsm2 yubiHsm2 =
+        new YubiHsm2(
+            yubiHsmSimulatorBinArgs(),
+            Optional.of(envVar()),
+            "http://localhost:12345",
+            SECONDARY_AUTH_KEY,
+            SECONDARY_PASSWORD,
+            Optional.empty(),
+            Optional.empty());
+
+    for (int i = 1; i <= 10; i++) {
+      final String key = yubiHsm2.fetchKey((short) i);
+      assertThat(key).isEqualTo(expectedKeys.get(i - 1));
+    }
+  }
+
+  @Test
+  public void errorIsReportedIfOpaqueObjectIdDoesNotExist() {
+    final YubiHsm2 yubiHsm2 =
+        new YubiHsm2(
+            yubiHsmSimulatorBinArgs(),
+            Optional.of(envVar()),
+            "http://localhost:12345",
+            SECONDARY_AUTH_KEY,
+            SECONDARY_PASSWORD,
+            Optional.empty(),
+            Optional.empty());
+
+    assertThatExceptionOfType(YubiHsmException.class)
+        .isThrownBy(() -> yubiHsm2.fetchKey((short) 11))
+        .withMessageContaining("Unable to get opaque object");
+  }
+
+  @Test
+  public void errorIsReportedIfInvalidAuthKeyIsUsed() {
+    final YubiHsm2 yubiHsm2 =
+        new YubiHsm2(
+            yubiHsmSimulatorBinArgs(),
+            Optional.of(envVar()),
+            "http://localhost:12345",
+            (short) 4,
+            SECONDARY_PASSWORD,
+            Optional.empty(),
+            Optional.empty());
+
+    assertThatExceptionOfType(YubiHsmException.class)
+        .isThrownBy(() -> yubiHsm2.fetchKey((short) 11))
+        .withMessage("Unable to fetch data from YubiHSM: Failed to open Session");
+  }
+
+  @Test
+  public void errorIsReportedIfInvalidAuthPasswordIsUsed() {
+    final YubiHsm2 yubiHsm2 =
+        new YubiHsm2(
+            yubiHsmSimulatorBinArgs(),
+            Optional.of(envVar()),
+            "http://localhost:12345",
+            SECONDARY_AUTH_KEY,
+            DEFAULT_PASSWORD,
+            Optional.empty(),
+            Optional.empty());
+
+    assertThatExceptionOfType(YubiHsmException.class)
+        .isThrownBy(() -> yubiHsm2.fetchKey((short) 11))
+        .withMessage("Unable to fetch data from YubiHSM: Failed to open Session");
+  }
+
+  private static void addOpaqueData() throws IOException, TimeoutException, InterruptedException {
+    for (int i = 1; i <= 10; i++) {
+      final BLSKeyPair blsKeyPair = BLSKeyPair.random(i);
```

Comment:
the storage doesn't need to be explicitly BLS oriented - it should be just random strings.

Reply:
You are correct.

Reply:
done.